File: Python/Linked_List/160.相交链表.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
	def getIntersectionNode(self, headA, headB):
		"""
		:type head1, head1: ListNode
		:rtype: ListNode
		"""
		node = headA
		lenA = 0
		while node != None:
			node = node.next
			lenA += 1
		node = headB
		lenB = 0
		while node != None:
			node = node.next
			lenB += 1

		if lenA >= lenB:
			start = headA
			node = headB
			diff = lenA - lenB
		else:
			start = headB
			node = headA
			diff = lenB - lenA

		for i in range(diff):
			start = start.next

		while start != None:
			if start == node:
				return start
			start = start.next
			node = node.next

			

File: Go/LinkedList/160-相交链表.go
package LinkedList

// 差值法，现将长的链表移到和短的链表一样长的地方，然后开始遍历判断
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	lenA, lenB := 0, 0
	a := headA
	b := headB
	for a.Next != nil {
		lenA++
		a = a.Next
	}
	for b.Next != nil {
		lenB++
		b = b.Next
	}
	// 根据大小重新复制，避免代码冗余
	if lenB > lenA {
		lenA, lenB = lenB, lenA
		headA, headB = headB, headA
	}
	for i:=0; i<lenA-lenB; i++ {
		headA = headA.Next
	}
	for headA != nil && headB != nil {
		if headA == headB {
			return headA
		}
		headA = headA.Next
		headB = headB.Next
	}
	return nil
}